Apromore-ProcessMining-Collection/prom-bpmn/src/main/java/org/apromore/processmining/models/graphbased/directed/bpmn/ModelChecker.java

```diff
@@ -0,0 +1,210 @@
+     * @param targetNodes a list of nodes in the model which have incoming edges.
+     */
+    private void checkStartEvent(Event event, List<BPMNNode> sourceNodes, List<BPMNNode> targetNodes) {
+        if (Collections.frequency(sourceNodes, event) > 1) {
+            // > 1 outgoing arc - start event appears more than once as a flow source
+            addError(ModelCheckResult.START_MULTIPLE_OUTGOING_ARCS_CODE, "START_EVENT");
+        } else if (!sourceNodes.contains(event)) {
+            // No outgoing arc - start event does not appear as a flow source
+            addError(ModelCheckResult.START_NO_OUTGOING_ARC_CODE, "START_EVENT");
+        }
+
+        if (targetNodes.contains(event)) {
+            // Any number of incoming arcs - start event appears as a flow target
+            addError(ModelCheckResult.START_INCOMING_ARCS_CODE, "START_EVENT");
+        }
+    }
+
+    /**
+     * Finds errors in an end event.
+     * @param event the end event to find errors in.
+     * @param sourceNodes a list of nodes in the model which have outgoing edges.
+     * @param targetNodes a list of nodes in the model which have incoming edges.
+     */
+    private void checkEndEvent(Event event, List<BPMNNode> sourceNodes, List<BPMNNode> targetNodes) {
+        if (Collections.frequency(targetNodes, event) > 1) {
+            // > 1 incoming arc - end event appears more than once as a flow target
+            addError(ModelCheckResult.END_MULTIPLE_INCOMING_ARCS_CODE, "END_EVENT");
+        } else if (!targetNodes.contains(event)) {
+            // No incoming arc - end event does not appear as a flow target
+            addError(ModelCheckResult.END_NO_INCOMING_ARC_CODE, "END_EVENT");
+        }
+
+        if (sourceNodes.contains(event)) {
+            // Any number of outgoing arcs - end event appears as a flow source
+            addError(ModelCheckResult.END_OUTGOING_ARCS_CODE, "END_EVENT");
+        }
+    }
+
+    /**
+     * Finds errors in a gateway.
+     * @param gateway the gateway to find errors in.
+     * @param sourceNodes a list of nodes in the model which have outgoing edges.
+     * @param targetNodes a list of nodes in the model which have incoming edges.
+     */
+    private void checkGateway(Gateway gateway, List<BPMNNode> sourceNodes, List<BPMNNode> targetNodes) {
+        if (!sourceNodes.contains(gateway) || !targetNodes.contains(gateway)) {
+            // missing incoming or outgoing arc - gateway does not appear as a flow source or target
+            addError(ModelCheckResult.GATE_MISSING_ARCS_CODE, gateway.getId().toString());
+        }
+    }
+
+    /**
+     * Add an error to the error map.
+     * @param errorCode the error code.
+     * @param element the name or id of the element with an error.
+     */
+    private void addError(int errorCode, String element) {
+        List<String> elementList = errorMap.get(errorCode);
+        if (elementList == null) {
+            elementList = new ArrayList<>();
+        }
```

Comment:
can be replaced with one line: errorMap.getOrDefault(errorCode, new ArrayList<>())